libmamba/src/core/download.cpp

```diff
@@ -126,33 +126,47 @@ namespace mamba
      * DownloadAttempt implementation *
      **********************************/
 
-    DownloadAttempt::DownloadAttempt(const DownloadRequest& request)
-        : p_request(&request)
-        , p_stream(nullptr)
+    DownloadAttempt::DownloadAttempt(
+        CURLHandle& handle,
+        const DownloadRequest& request,
+        CURLMultiHandle& downloader,
+        const Context& context,
+        on_success_callback success,
+        on_failure_callback error
+    )
+        : p_impl(std::make_unique<
+                 Impl>(handle, request, downloader, context, std::move(success), std::move(error)))
+    {
+    }
+
+    auto DownloadAttempt::create_completion_function() -> completion_function
     {
-        m_retry_wait_seconds = std::size_t(0);
+        return [impl = p_impl.get()](CURLMultiHandle& handle, CURLcode code)
```

Comment:
If you find issues later with this lambda being called after `impl` have been destroyed, consider using task_synchronizer to guarantee that this does not ever happen.